### PR TITLE
python312Packages.pytorch3d: init at 0.7.8

### PR DESCRIPTION
**Pull request by SomeoneSerge**
Bootstrapped with nix-init and built with cudaSupport=true. Yet to run any tests beyond pythonImportsCheck

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/pytorch3d/default.nix [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  nativeBuildInputs = lib.optionals cudaSupport [ cudaPackages.cuda_nvcc ];
+  build-system = [
+    setuptools
+    wheel
+  ];
+  dependencies = [
+    torch
+    iopath
+  ];
+  buildInputs = [ (lib.getOutput "cxxdev" torch) ];
```
Comment (pbsds) [inline pkgs/development/python-modules/pytorch3d/default.nix:37]: TIL

File: pkgs/development/python-modules/pytorch3d/default.nix [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    owner = "facebookresearch";
+    repo = "pytorch3d";
+    rev = "V${version}";
+    hash = "sha256-DEEWWfjwjuXGc0WQInDTmtnWSIDUifyByxdg7hpdHlo=";
+  };
+
+  nativeBuildInputs = lib.optionals cudaSupport [ cudaPackages.cuda_nvcc ];
+  build-system = [
+    setuptools
+    wheel
```
Comment (pbsds) [inline pkgs/development/python-modules/pytorch3d/default.nix:31]:
```suggestion
```

**Review by pbsds — COMMENTED** [review]
I don't have a nixos-unstable cuda host to test on, but i based this on `saxpy` which you showed me yesterday:

```nix
  passthru.gpuCheck = pytorch3d.overrideAttrs (_: {
    requiredSystemFeatures = [ "cuda" ];
    doInstallCheck = true;
    postInstallCheck = ''
      python -c '
        import pytorch3d.transforms as p3dt

        M = p3dt.random_rotations(n=10, device="cuda")
        assert "cuda" in M.device.type
        angles = p3dt.matrix_to_euler_angles(M, "XYZ")
        assert "cuda" in angles.device.type
        assert angles.shape == (10, 3), angles.shape
        print(angles)
      '
    '';
  });
```

I believe you also had a cuda python test runner, but i didn't find it with a grep.

File: pkgs/development/python-modules/pytorch3d/default.nix (hunk 1 above)
Comment (SomeoneSerge) [reply pkgs/development/python-modules/pytorch3d/default.nix:37]: Hopefully we can remove this once we switch to eval-time propagation of python `dependencies`...

File: pkgs/development/python-modules/pytorch3d/default.nix (hunk 2 above)
Comment (SomeoneSerge) [reply pkgs/development/python-modules/pytorch3d/default.nix:31]: H'm. It came from nix-init, it's probably declared in the toml?

File: pkgs/development/python-modules/pytorch3d/default.nix [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    cudaPackages.writeGpuTestPython { libraries = ps: [ ps.pytorch3d ]; }
+      ''
+        import pytorch3d.transforms as p3dt
+
+        M = p3dt.random_rotations(n=10, device="cuda")
+        assert "cuda" in M.device.type
+        angles = p3dt.matrix_to_euler_angles(M, "XYZ")
+        assert "cuda" in angles.device.type
+        assert angles.shape == (10, 3), angles.shape
+        print(angles)
```
Comment (SomeoneSerge) [inline pkgs/development/python-modules/pytorch3d/default.nix:59]: So the test worked even without `FORCE_CUDA=1`. Pytorch3d must be falling back to pure python/pure pytorch mode without fused kernels or something like that?

**Pull request merged**